backend/controllers/searchWithFilter.js

```diff
@@ -0,0 +1,107 @@
+const querystring = require('querystring')
+
+const searchFactory = (
+  paginationService,
+  getAssignableGroupsOrPrisonsApi,
+  getSearchableRolesApi,
+  searchApi,
+  pagingApi,
+  searchUrl,
+  maintainUrl,
+  searchTitle,
+  dpsSearch,
+  allowDownload,
+) => {
+  return async (req, res) => {
+    const [groupOrPrisonDropdownValues, searchableRoles] = await Promise.all([
+      getAssignableGroupsOrPrisonsApi(res.locals),
+      getSearchableRolesApi(res.locals),
+    ])
+    const roleDropdownValues = searchableRoles.map((r) => ({
+      text: r.roleName,
+      value: r.roleCode,
+    }))
+    const showGroupOrPrisonDropdown = Boolean(res.locals?.user?.maintainAccessAdmin || !dpsSearch)
+
+    const currentFilter = parseFilter(req.query)
+
+    const { offset } = req.query
+
+    const pageSize = 20
+    const pageOffset = (offset && parseInt(offset, 10)) || 0
+
+    const groupCode = currentFilter.groupCode && currentFilter.groupCode[0]
+    const roleCode = (currentFilter.roleCode && currentFilter.roleCode[0]) ?? ''
```

Comment:
A few funnies here to do with how the service builds prison API request parameters - one uses query stringy that is happy with 'undefined' and another builds with strings which isn't - hence falling back to empty string for some parameters used for DPS non-admin search